### PR TITLE
Enable Firefox, Fennec and Thunderbird localization

### PR DESCRIPTION
**Pull request by mathjazz**
I'll try to avoid so big PRs in the future.

**Enable Firefox, Fennec and Thunderbird localization**
* Support for .inc file format
* Support for standard .ini file format instead of fake webL10n one
* Mozilla-specific code needed for translating Firefox (a85ee5c, d2fa738, 74ea416)
* Wider resource menu
* Update SIlme

**Other new features**
* Highlight current file in the menu
* Support for .l20n file format
* Update translate-toolkit

**Bugfixes**
* Unify resource order
* Do not show obsolete files in resource menu
* Remove ununnecessary unsaved changes warning 
* Do not escape placeables when inserting into translation area

### CODE REVIEW

**Comment by mathjazz** [comment]
@Osmose, r?

File: pontoon/administration/files.py [hunk 1]
```diff
@@ -53,8 +66,8 @@ def split_comments_mine(
             cls.split_entities(
                 text, object, code=code, pointer=pointer, end=st0)
         groups = match.groups()
-        comment = silme.core.structure.Comment(
-            match.group(0)[1:].replace('\n#', '\n'))
+        comment = Comment()
+        comment.add(match.group(0)[1:].replace('\n#', '\n'))
```
Comment (Osmose) [inline pontoon/administration/files.py]: So after looking at silme's code and briefly at [this tutorial](https://wiki.mozilla.org/Silme:Tutorial:Concepts), this seems to be avoiding parsing entities out of comment nodes, so that the resulting L10nObject has just a single Comment node with the string contents inside of it. Is that correct?

If so, why bother removing this functionality?

File: pontoon/administration/files.py [hunk 2]
```diff
@@ -64,27 +77,56 @@ def split_comments_mine(
...
+        if end:
+            match = pattern.search(text, pointer, end)
+        else:
+            match = pattern.search(text, pointer)
+    if (not end or (end > pointer)) and len(text) > pointer:
+        cls.split_entities(text, object, pointer=pointer, end=end)
+
+IniParser.split_comments = split_comments_ini
+
+IniParser.patterns['entity'] = re.compile(
```
Comment (Osmose) [inline pontoon/administration/files.py]: A comment explaining what this changes and why the changes is necessary would be really helpful.

File: pontoon/administration/files.py [hunk 3]
```diff
@@ -64,27 +77,56 @@ def split_comments_mine(
...
+    '^[ \t]*([^#!\s\n][^=:\n]*?)[ \t]*[:=][ \t]*(.*?)(?<!\\\)(?=\n|\Z)',
+    re.S | re.M)
 
 
 def __repr__mine(self):
-        string = ''
```
Comment (Osmose) [inline pontoon/administration/files.py]: Why bother stripping out the `<comment>` stuff from silme?

File: pontoon/administration/files.py (hunk 1 above)
Comment (mathjazz) [inline pontoon/administration/files.py]: That's correct, I'm avoiding parsing comments, because it doesn't work as expected.

See:
https://bugzilla.mozilla.org/show_bug.cgi?id=1017849

File: pontoon/administration/files.py (hunk 2 above)
Comment (mathjazz) [inline pontoon/administration/files.py]: Same thing as above (bug 1017849).

I'm aware this is a hack and initially I thought it will be temporary. ;) I was hoping Silme will get fixed soon with my patch, but it didn't happen. I tried switching to [translate toolkit](http://toolkit.translatehouse.org/), but it also doesn't work entirely as expected, so I would need to patch it, too.

File: pontoon/administration/files.py (hunk 3 above)
Comment (mathjazz) [inline pontoon/administration/files.py]: We need a simple string, without any markup:
http://hg.mozilla.org/l10n/silme/file/b21c5151136e/lib/silme/core/structure.py#l260

Comment (Osmose) [inline pontoon/administration/files.py]: Why do you need that from `repr`? Where are you using it that that is getting in the way?

File: pontoon/administration/files.py (hunk 1 above)
Comment (Osmose) [inline pontoon/administration/files.py]: 1 year? Jeebus. We should just fork the code to Github, apply our fix, and then pull in that version. All this replicated code is kinda scary to me.

**Comment by Osmose** [comment]
Just as a note, a lot of my comments are addressing larger issues than what this PR aims to handle, so unless I specifically call something out as a blocker, assume that my comments are things that would be nice to fix but can be deferred if this PR needs to be merged ASAP.

Also as a note, this is an impressive chunk of work and the end result is awesome, nice work! Sorry if I'm a little slow in reviewing.

File: pontoon/administration/files.py (hunk 3 above)
Comment (Osmose) [inline pontoon/administration/files.py]: I see in https://github.com/mathjazz/pontoon/commit/a816e7d48383fbc33c1cb7fa7901bd48006ba2f3 that we're using this in `extract_silme`. If that's the only spot, I'd recommend avoiding monkey-patching the class and just using this code directly in that method. If we're the only consumer there's no reason to patch.

Also, a generator and `join` would be a slightly faster, more pythonic way of doing this:

```python
return ''.join(unicode(i) for i in self)
```

File: pontoon/administration/files.py (hunk 1 above)
Comment (mathjazz) [inline pontoon/administration/files.py]: https://bugzilla.mozilla.org/show_bug.cgi?id=1181604

File: pontoon/administration/files.py (hunk 3 above)
Comment (mathjazz) [inline pontoon/administration/files.py]: It's the only spot, but without touching the class, I would need to parse comments from " <comment:  Device :: Help>" and the likes. I think this should be part of https://bugzilla.mozilla.org/show_bug.cgi?id=1181604.

Code updated.

Comment (Osmose) [inline pontoon/administration/files.py]: Couldn't you just do this and avoid the extra comment stuff?

```python
comment = ''.join(unicode(i) for i in obj)
```

**Comment by Osmose** [comment]
Overall, nice work! This was a lot to review but it was really helpful in getting me to better understand the extraction and exporting process. 

The only blocker I found was that plurals model; let me know when you've updated with a fix and I'll give it a quick look. Otherwise this is an r+. :+1:

**Comment by mathjazz** [comment]
I fixed the plurals model and also found annoying regression that has to land ASAP.

**Comment by Osmose** [comment]
That `choices` issue is the only serious thing remaining. Great work!

**Comment by mathjazz** [comment]
I merged this before @Osmose gave r+ for the blocker part, because I wanted to deploy ASAP (this is very imporant PR). I failed though, got Internal Server Error due to missing requirements I guess:
https://pastebin.mozilla.org/8839344

I'd rather leave the deployment part to you, @Osmose.

**Comment by Osmose** [comment]
r+ on the code, nice work!
